pkg/controllers/nodeclaim/tagging/controller.go

```diff
@@ -0,0 +1,133 @@
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"knative.dev/pkg/logging"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/samber/lo"
+
+	"github.com/aws/karpenter/pkg/apis/v1beta1"
+	"github.com/aws/karpenter/pkg/providers/instance"
+	"github.com/aws/karpenter/pkg/utils"
+
+	corev1beta1 "github.com/aws/karpenter-core/pkg/apis/v1beta1"
+	"github.com/aws/karpenter-core/pkg/cloudprovider"
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+)
+
+const (
+	TagNodeClaim = corev1beta1.Group + "/nodeclaim"
+	TagName      = "Name"
+)
+
+type Controller struct {
+	kubeClient       client.Client
+	instanceProvider *instance.Provider
+}
+
+func NewController(kubeClient client.Client, instanceProvider *instance.Provider) corecontroller.Controller {
+	return corecontroller.Typed[*corev1beta1.NodeClaim](kubeClient, &Controller{
+		kubeClient:       kubeClient,
+		instanceProvider: instanceProvider,
+	})
+}
+
+func (c *Controller) Name() string {
+	return "nodeclaim.tagging"
+}
+
+func (c *Controller) Reconcile(ctx context.Context, nodeClaim *corev1beta1.NodeClaim) (reconcile.Result, error) {
+	storedNodeClaim := nodeClaim.DeepCopy()
+
+	if !isTaggable(nodeClaim) {
+		return reconcile.Result{}, nil
+	}
+
+	id, err := utils.ParseInstanceID(nodeClaim.Status.ProviderID)
+	if err != nil {
+		// We don't throw an error here since we don't want to retry until the ProviderID has been updated.
+		logging.FromContext(ctx).Errorf("failed to parse instance ID, %w", err)
+		return reconcile.Result{}, nil
+	}
+
+	err = c.tagInstance(ctx, nodeClaim, id)
+	if err != nil {
+		return reconcile.Result{}, cloudprovider.IgnoreNodeClaimNotFoundError(err)
+	}
+
+	nodeClaim.Annotations = lo.Assign(nodeClaim.Annotations, map[string]string{v1beta1.AnnotationInstanceTagged: "true"})
+	if !equality.Semantic.DeepEqual(nodeClaim, storedNodeClaim) {
+		if err := c.kubeClient.Patch(ctx, nodeClaim, client.MergeFrom(storedNodeClaim)); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(err)
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Builder(_ context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(
+		controllerruntime.
+			NewControllerManagedBy(m).
+			For(&corev1beta1.NodeClaim{}).
+			WithEventFilter(predicate.NewPredicateFuncs(func(o client.Object) bool {
+				return isTaggable(o.(*corev1beta1.NodeClaim))
+			})),
+	)
+}
+
+func (c *Controller) tagInstance(ctx context.Context, nc *corev1beta1.NodeClaim, id string) error {
+	tags := map[string]string{
+		TagName:      nc.Status.NodeName,
+		TagNodeClaim: nc.Name,
+	}
+
+	// Remove tags which have been already populated
+	instance, err := c.instanceProvider.Get(ctx, id)
+	if err != nil {
+		return err
```

Comment:
```suggestion
		return fmt.Errorf("tagging nodeclaim, %w", err)
```

nit: contextualize the error